gbpservice/neutron/services/grouppolicy/drivers/oc_resource_mapping.py

```diff
@@ -0,0 +1,271 @@
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import netaddr
+
+from neutron.common import log
+from neutron.openstack.common import jsonutils
+from neutron.openstack.common import log as logging
+from neutron.plugins.common import constants as pconst
+from neutron.common import exceptions as n_exc
+from neutron.extensions import securitygroup as ext_sg
+from oslo.config import cfg
+
+from gbpservice.neutron.services.grouppolicy.common import constants as gconst
+from gbpservice.neutron.services.grouppolicy.common import exceptions as exc
+from gbpservice.neutron.extensions import group_policy as gp_ext
+
+from gbpservice.neutron.services.grouppolicy.drivers import resource_mapping
+
+LOG = logging.getLogger(__name__)
+
+
+class OCResourceMappingDriver(resource_mapping.ResourceMappingDriver):
+    """Resource Mapping driver for Group Policy plugin.
+
+    This driver implements group policy semantics by mapping group
+    policy resources to various other neutron resources.
+    """
+
+    @log.log
+    def create_policy_target_group_postcommit(self, context):
+        subnets = context.current['subnets']
+        if subnets:
+            l2p_id = context.current['l2_policy_id']
+            l2p = context._plugin.get_l2_policy(context._plugin_context,
+                                                l2p_id)
+            l3p_id = l2p['l3_policy_id']
+            l3p = context._plugin.get_l3_policy(context._plugin_context,
+                                                l3p_id)
+            router_id = l3p['routers'][0] if l3p['routers'] else None
+            for subnet_id in subnets:
+                self._use_explicit_subnet(context._plugin_context, subnet_id,
+                                          router_id)
+        else:
+            self._use_implicit_subnet(context)
+        self._handle_network_service_policy(context)
+        self._handle_policy_rule_sets(context)
+        self._update_default_security_group(context._plugin_context,
+                                            context.current['id'],
+                                            context.current['tenant_id'],
+                                            context.current['subnets'])
+        if (self._is_firewall_in_sc_spec(context, ptg_id=context.current['id'])
+            and self._is_ptg_provider(context, context.current['id'])):
+            self._create_service_vm_sg(context,
+                                       context.current['tenant_id'],
+                                       context.current['id'])
```

Comment:
What happens if there is no firewall? How do wehandle that case?  Should we raise an exception?